// @genkit-ai/tools-common/lib/cjs/src/types/parts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultipartToolResponseSchema = exports.PartSchema = exports.ResourcePartSchema = exports.CustomPartSchema = exports.DataPartSchema = exports.ToolResponsePartSchema = exports.ToolResponseSchema = exports.ToolRequestPartSchema = exports.ToolRequestSchema = exports.MediaPartSchema = exports.MediaSchema = exports.ReasoningPartSchema = exports.TextPartSchema = void 0;
const zod_1 = require("zod");
const EmptyPartSchema = zod_1.z.object({
    text: zod_1.z.never().optional(),
    media: zod_1.z.never().optional(),
    toolRequest: zod_1.z.never().optional(),
    toolResponse: zod_1.z.never().optional(),
    data: zod_1.z.unknown().optional(),
    metadata: zod_1.z.record(zod_1.z.unknown()).optional(),
    custom: zod_1.z.record(zod_1.z.unknown()).optional(),
    reasoning: zod_1.z.never().optional(),
    resource: zod_1.z.never().optional(),
});
exports.TextPartSchema = EmptyPartSchema.extend({
    text: zod_1.z.string(),
});
exports.ReasoningPartSchema = EmptyPartSchema.extend({
    reasoning: zod_1.z.string(),
});
exports.MediaSchema = zod_1.z.object({
    contentType: zod_1.z.string().optional(),
    url: zod_1.z.string(),
});
exports.MediaPartSchema = EmptyPartSchema.extend({
    media: exports.MediaSchema,
});
exports.ToolRequestSchema = zod_1.z.object({
    ref: zod_1.z.string().optional(),
    name: zod_1.z.string(),
    input: zod_1.z.unknown().optional(),
    partial: zod_1.z.boolean().optional(),
});
exports.ToolRequestPartSchema = EmptyPartSchema.extend({
    toolRequest: exports.ToolRequestSchema,
});
const ToolResponseSchemaBase = zod_1.z.object({
    ref: zod_1.z.string().optional(),
    name: zod_1.z.string(),
    output: zod_1.z.unknown().optional(),
});
exports.ToolResponseSchema = ToolResponseSchemaBase.extend({
    content: zod_1.z.array(zod_1.z.any()).optional(),
});
exports.ToolResponsePartSchema = EmptyPartSchema.extend({
    toolResponse: exports.ToolResponseSchema,
});
exports.DataPartSchema = EmptyPartSchema.extend({
    data: zod_1.z.unknown(),
});
exports.CustomPartSchema = EmptyPartSchema.extend({
    custom: zod_1.z.record(zod_1.z.any()),
});
exports.ResourcePartSchema = EmptyPartSchema.extend({
    resource: zod_1.z.object({
        uri: zod_1.z.string(),
    }),
});
exports.PartSchema = zod_1.z.union([
    exports.TextPartSchema,
    exports.MediaPartSchema,
    exports.ToolRequestPartSchema,
    exports.ToolResponsePartSchema,
    exports.DataPartSchema,
    exports.CustomPartSchema,
    exports.ReasoningPartSchema,
    exports.ResourcePartSchema,
]);
exports.MultipartToolResponseSchema = zod_1.z.object({
    output: zod_1.z.unknown().optional(),
    content: zod_1.z.array(exports.PartSchema).optional(),
});
