// @genkit-ai/tools-common/lib/cjs/src/manager/manager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeManager = exports.GENKIT_REFLECTION_API_SPEC_VERSION = void 0;
const axios_1 = __importDefault(require("axios"));
const chokidar_1 = __importDefault(require("chokidar"));
const events_1 = __importDefault(require("events"));
const fsSync = __importStar(require("fs"));
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const action_1 = require("../types/action");
const apis = __importStar(require("../types/apis"));
const logger_1 = require("../utils/logger");
const utils_1 = require("../utils/utils");
const types_1 = require("./types");
const STREAM_DELIMITER = '\n';
const HEALTH_CHECK_INTERVAL = 5000;
exports.GENKIT_REFLECTION_API_SPEC_VERSION = 1;
class RuntimeManager {
    telemetryServerUrl;
    manageHealth;
    projectRoot;
    processManager;
    disableRealtimeTelemetry;
    filenameToRuntimeMap = {};
    filenameToDevUiMap = {};
    idToFileMap = {};
    eventEmitter = new events_1.default();
    watchers = [];
    healthCheckInterval;
    constructor(telemetryServerUrl, manageHealth, projectRoot, processManager, disableRealtimeTelemetry) {
        this.telemetryServerUrl = telemetryServerUrl;
        this.manageHealth = manageHealth;
        this.projectRoot = projectRoot;
        this.processManager = processManager;
        this.disableRealtimeTelemetry = disableRealtimeTelemetry ?? false;
    }
    static async create(options) {
        const manager = new RuntimeManager(options.telemetryServerUrl, options.manageHealth ?? true, options.projectRoot, options.processManager, options.disableRealtimeTelemetry);
        await manager.setupRuntimesWatcher();
        await manager.setupDevUiWatcher();
        if (manager.manageHealth) {
            manager.healthCheckInterval = setInterval(async () => await manager.performHealthChecks(), HEALTH_CHECK_INTERVAL);
        }
        return manager;
    }
    async stop() {
        if (this.healthCheckInterval) {
            clearInterval(this.healthCheckInterval);
        }
        await Promise.all(this.watchers.map((watcher) => watcher.close()));
        if (this.processManager) {
            await this.processManager.kill();
        }
    }
    listRuntimes() {
        return Object.values(this.filenameToRuntimeMap);
    }
    getRuntimeById(id) {
        const fileName = this.idToFileMap[id];
        return fileName ? this.filenameToRuntimeMap[fileName] : undefined;
    }
    getMostRecentRuntime() {
        const runtimes = Object.values(this.filenameToRuntimeMap);
        return runtimes.length === 0
            ? undefined
            : runtimes.reduce((a, b) => new Date(a.timestamp) > new Date(b.timestamp) ? a : b);
    }
    getMostRecentDevUI() {
        const toolsInfo = Object.values(this.filenameToDevUiMap);
        return toolsInfo.length === 0
            ? undefined
            : toolsInfo.reduce((a, b) => new Date(a.timestamp) > new Date(b.timestamp) ? a : b);
    }
    onRuntimeEvent(listener) {
        Object.values(types_1.RuntimeEvent).forEach((event) => this.eventEmitter.on(event, (rt) => listener(event, rt)));
    }
    async listActions(input) {
        const runtime = input?.runtimeId
            ? this.getRuntimeById(input.runtimeId)
            : this.getMostRecentRuntime();
        if (!runtime) {
            throw new Error(input?.runtimeId
                ? `No runtime found with ID ${input.runtimeId}.`
                : 'No runtimes found. Make sure your app is running using `genkit start -- ...`. See getting started documentation.');
        }
        const response = await axios_1.default
            .get(`${runtime.reflectionServerUrl}/api/actions`)
            .catch((err) => this.httpErrorHandler(err, 'Error listing actions.'));
        return response.data;
    }
    async runAction(input, streamingCallback, onTraceId) {
        const runtime = input.runtimeId
            ? this.getRuntimeById(input.runtimeId)
            : this.getMostRecentRuntime();
        if (!runtime) {
            throw new Error(input.runtimeId
                ? `No runtime found with ID ${input.runtimeId}.`
                : 'No runtimes found. Make sure your app is running using `genkit start -- ...`. See getting started documentation.');
        }
        if (streamingCallback) {
            const response = await axios_1.default
                .post(`${runtime.reflectionServerUrl}/api/runAction?stream=true`, input, {
                headers: {
                    'Content-Type': 'application/json',
                },
                responseType: 'stream',
            })
                .catch((err) => this.handleStreamError(err, `Error running action key='${input.key}'.`));
            let genkitVersion;
            if (response.headers['x-genkit-version']) {
                genkitVersion = response.headers['x-genkit-version'];
            }
            const traceId = response.headers['x-genkit-trace-id'];
            if (traceId && onTraceId) {
                onTraceId(traceId);
            }
            const stream = response.data;
            let buffer = '';
            stream.on('data', (data) => {
                buffer += data;
                while (buffer.includes(STREAM_DELIMITER)) {
                    try {
                        streamingCallback(JSON.parse(buffer.substring(0, buffer.indexOf(STREAM_DELIMITER))));
                        buffer = buffer.substring(buffer.indexOf(STREAM_DELIMITER) + STREAM_DELIMITER.length);
                    }
                    catch (err) {
                        logger_1.logger.error(`Bad stream: ${err}`);
                        break;
                    }
                }
            });
            let resolver;
            let rejecter;
            const promise = new Promise((resolve, reject) => {
                resolver = resolve;
                rejecter = reject;
            });
            stream.on('end', () => {
                const parsedBuffer = JSON.parse(buffer);
                if (parsedBuffer.error) {
                    const err = new types_1.GenkitToolsError(`Error running action key='${input.key}'.`);
                    err.data = {
                        ...parsedBuffer.error,
                        stack: (parsedBuffer.error?.details).stack,
                        data: {
                            genkitErrorMessage: parsedBuffer.error?.message,
                            genkitErrorDetails: parsedBuffer.error?.details,
                        },
                    };
                    rejecter(err);
                    return;
                }
                const actionResponse = action_1.RunActionResponseSchema.parse(parsedBuffer);
                if (genkitVersion) {
                    actionResponse.genkitVersion = genkitVersion;
                }
                resolver(actionResponse);
            });
            stream.on('error', (err) => {
                rejecter(err);
            });
            return promise;
        }
        else {
            const response = await axios_1.default
                .post(`${runtime.reflectionServerUrl}/api/runAction`, input, {
                headers: {
                    'Content-Type': 'application/json',
                },
                responseType: 'stream',
            })
                .catch((err) => this.handleStreamError(err, `Error running action key='${input.key}'.`));
            const traceId = response.headers['x-genkit-trace-id'];
            if (traceId && onTraceId) {
                onTraceId(traceId);
            }
            return new Promise((resolve, reject) => {
                let buffer = '';
                response.data.on('data', (chunk) => {
                    buffer += chunk.toString();
                });
                response.data.on('end', () => {
                    try {
                        const responseData = JSON.parse(buffer);
                        if (responseData.error) {
                            const err = new types_1.GenkitToolsError(`Error running action key='${input.key}'.`);
                            err.data = {
                                ...responseData.error,
                                stack: (responseData.error?.details).stack,
                                data: {
                                    genkitErrorMessage: responseData.error?.message,
                                    genkitErrorDetails: responseData.error?.details,
                                },
                            };
                            reject(err);
                            return;
                        }
                        if (!responseData.telemetry && traceId) {
                            responseData.telemetry = { traceId: traceId };
                        }
                        const parsed = action_1.RunActionResponseSchema.parse(responseData);
                        if (response.headers['x-genkit-version']) {
                            parsed.genkitVersion = response.headers['x-genkit-version'];
                        }
                        resolve(parsed);
                    }
                    catch (err) {
                        reject(new types_1.GenkitToolsError(`Failed to parse response: ${err}`));
                    }
                });
                response.data.on('error', (err) => {
                    reject(err);
                });
            });
        }
    }
    async cancelAction(input) {
        const runtime = input.runtimeId
            ? this.getRuntimeById(input.runtimeId)
            : this.getMostRecentRuntime();
        if (!runtime) {
            throw new Error(input.runtimeId
                ? `No runtime found with ID ${input.runtimeId}.`
                : 'No runtimes found. Make sure your app is running.');
        }
        try {
            const response = await axios_1.default.post(`${runtime.reflectionServerUrl}/api/cancelAction`, { traceId: input.traceId }, {
                headers: {
                    'Content-Type': 'application/json',
                },
            });
            return response.data;
        }
        catch (err) {
            const axiosError = err;
            if (axiosError.response?.status === 404) {
                const error = new types_1.GenkitToolsError('Action not found or already completed');
                error.data = {
                    message: 'Action not found or already completed',
                };
                error.data.statusCode = 404;
                throw error;
            }
            throw this.httpErrorHandler(axiosError);
        }
    }
    async listTraces(input) {
        const { limit, continuationToken, filter } = input;
        let query = '';
        if (limit) {
            query += `limit=${limit}`;
        }
        if (continuationToken) {
            if (query !== '') {
                query += '&';
            }
            query += `continuationToken=${continuationToken}`;
        }
        if (filter) {
            if (query !== '') {
                query += '&';
            }
            query += `filter=${encodeURI(JSON.stringify(filter))}`;
        }
        const response = await axios_1.default
            .get(`${this.telemetryServerUrl}/api/traces?${query}`)
            .catch((err) => this.httpErrorHandler(err, `Error listing traces for query='${query}'.`));
        return apis.ListTracesResponseSchema.parse(response.data);
    }
    async getTrace(input) {
        const { traceId } = input;
        const response = await axios_1.default
            .get(`${this.telemetryServerUrl}/api/traces/${traceId}`)
            .catch((err) => this.httpErrorHandler(err, `Error getting trace for traceId='${traceId}'`));
        return response.data;
    }
    async streamTrace(input, streamingCallback) {
        const { traceId } = input;
        if (!this.telemetryServerUrl) {
            throw new Error('Telemetry server URL not configured. Cannot stream trace updates.');
        }
        const response = await axios_1.default
            .get(`${this.telemetryServerUrl}/api/traces/${traceId}/stream`, {
            headers: {
                Accept: 'text/event-stream',
            },
            responseType: 'stream',
        })
            .catch((err) => this.httpErrorHandler(err, `Error streaming trace for traceId='${traceId}'`));
        const stream = response.data;
        let buffer = '';
        return new Promise((resolve, reject) => {
            stream.on('data', (chunk) => {
                buffer += chunk.toString();
                while (buffer.includes('\n\n')) {
                    const messageEnd = buffer.indexOf('\n\n');
                    const message = buffer.substring(0, messageEnd).trim();
                    buffer = buffer.substring(messageEnd + 2);
                    if (!message) {
                        continue;
                    }
                    try {
                        const jsonData = message.startsWith('data: ')
                            ? message.slice(6)
                            : message;
                        const parsed = JSON.parse(jsonData);
                        streamingCallback(parsed);
                    }
                    catch (err) {
                        logger_1.logger.error(`Error parsing stream data: ${err}`);
                    }
                }
            });
            stream.on('end', () => {
                resolve();
            });
            stream.on('error', (err) => {
                logger_1.logger.error(`Stream error for traceId='${traceId}': ${err}`);
                reject(err);
            });
        });
    }
    async addTrace(input) {
        await axios_1.default
            .post(`${this.telemetryServerUrl}/api/traces/`, input)
            .catch((err) => this.httpErrorHandler(err, 'Error writing trace to store.'));
    }
    async notifyRuntime(runtime) {
        try {
            await axios_1.default.post(`${runtime.reflectionServerUrl}/api/notify`, {
                telemetryServerUrl: this.telemetryServerUrl,
                reflectionApiSpecVersion: exports.GENKIT_REFLECTION_API_SPEC_VERSION,
            });
        }
        catch (error) {
            logger_1.logger.error(`Failed to notify runtime ${runtime.id}: ${error}`);
        }
    }
    async setupRuntimesWatcher() {
        try {
            const runtimesDir = await (0, utils_1.findRuntimesDir)(this.projectRoot);
            await promises_1.default.mkdir(runtimesDir, { recursive: true });
            const watcher = chokidar_1.default.watch(runtimesDir, {
                persistent: true,
                ignoreInitial: false,
            });
            this.watchers.push(watcher);
            watcher.on('add', (filePath) => this.handleNewRuntime(filePath));
            if (this.manageHealth) {
                watcher.on('unlink', (filePath) => this.handleRemovedRuntime(filePath));
            }
            for (const runtime of await promises_1.default.readdir(runtimesDir)) {
                await this.handleNewRuntime(path_1.default.resolve(runtimesDir, runtime));
            }
        }
        catch (error) {
            logger_1.logger.error('Failed to set up runtimes watcher:', error);
        }
    }
    async setupDevUiWatcher() {
        try {
            const serversDir = await (0, utils_1.findServersDir)(this.projectRoot);
            await promises_1.default.mkdir(serversDir, { recursive: true });
            const watcher = chokidar_1.default.watch(serversDir, {
                persistent: true,
                ignoreInitial: false,
            });
            this.watchers.push(watcher);
            watcher.on('add', (filePath) => this.handleNewDevUi(filePath));
            if (this.manageHealth) {
                watcher.on('unlink', (filePath) => this.handleRemovedDevUi(filePath));
            }
            for (const toolsInfo of await promises_1.default.readdir(serversDir)) {
                await this.handleNewDevUi(path_1.default.resolve(serversDir, toolsInfo));
            }
        }
        catch (error) {
            logger_1.logger.error('Failed to set up tools server watcher:', error);
        }
    }
    async handleNewDevUi(filePath) {
        try {
            if (!fsSync.existsSync(filePath)) {
                return;
            }
            const { content, toolsInfo } = await (0, utils_1.retriable)(async () => {
                const content = await promises_1.default.readFile(filePath, 'utf-8');
                const toolsInfo = JSON.parse(content);
                return { content, toolsInfo };
            }, { maxRetries: 10, delayMs: 500 });
            if ((0, utils_1.isValidDevToolsInfo)(toolsInfo)) {
                const fileName = path_1.default.basename(filePath);
                if (await (0, utils_1.checkServerHealth)(toolsInfo.url)) {
                    this.filenameToDevUiMap[fileName] = toolsInfo;
                }
                else {
                    logger_1.logger.debug('Found an unhealthy tools config file', fileName);
                    await (0, utils_1.removeToolsInfoFile)(fileName, this.projectRoot);
                }
            }
            else {
                logger_1.logger.error(`Unexpected file in the servers directory: ${content}`);
            }
        }
        catch (error) {
            logger_1.logger.error('Error reading tools config', error);
            return undefined;
        }
    }
    handleRemovedDevUi(filePath) {
        const fileName = path_1.default.basename(filePath);
        if (fileName in this.filenameToDevUiMap) {
            const toolsInfo = this.filenameToDevUiMap[fileName];
            delete this.filenameToDevUiMap[fileName];
            logger_1.logger.debug(`Removed Dev UI with url ${toolsInfo.url}.`);
        }
    }
    async handleNewRuntime(filePath) {
        try {
            if (!fsSync.existsSync(filePath)) {
                return;
            }
            const { content, runtimeInfo } = await (0, utils_1.retriable)(async () => {
                const content = await promises_1.default.readFile(filePath, 'utf-8');
                const runtimeInfo = JSON.parse(content);
                runtimeInfo.projectName = (0, utils_1.projectNameFromGenkitFilePath)(filePath);
                return { content, runtimeInfo };
            }, { maxRetries: 10, delayMs: 500 });
            if (isValidRuntimeInfo(runtimeInfo)) {
                if (!runtimeInfo.name) {
                    runtimeInfo.name = runtimeInfo.id;
                }
                const fileName = path_1.default.basename(filePath);
                if (await (0, utils_1.checkServerHealth)(runtimeInfo.reflectionServerUrl, runtimeInfo.id)) {
                    if (runtimeInfo.reflectionApiSpecVersion !=
                        exports.GENKIT_REFLECTION_API_SPEC_VERSION) {
                        if (!runtimeInfo.reflectionApiSpecVersion ||
                            runtimeInfo.reflectionApiSpecVersion <
                                exports.GENKIT_REFLECTION_API_SPEC_VERSION) {
                            logger_1.logger.warn('Genkit CLI is newer than runtime library. Some feature may not be supported. ' +
                                'Consider upgrading your runtime library version (debug info: expected ' +
                                `${exports.GENKIT_REFLECTION_API_SPEC_VERSION}, got ${runtimeInfo.reflectionApiSpecVersion}).`);
                        }
                        else {
                            logger_1.logger.error('Genkit CLI version is outdated. Please update `genkit-cli` to the latest version.');
                            process.exit(1);
                        }
                    }
                    this.filenameToRuntimeMap[fileName] = runtimeInfo;
                    this.idToFileMap[runtimeInfo.id] = fileName;
                    this.eventEmitter.emit(types_1.RuntimeEvent.ADD, runtimeInfo);
                    await this.notifyRuntime(runtimeInfo);
                    logger_1.logger.debug(`Added runtime with ID ${runtimeInfo.id} at URL: ${runtimeInfo.reflectionServerUrl}`);
                }
                else {
                    await this.removeRuntime(fileName);
                }
            }
            else {
                logger_1.logger.error(`Unexpected file in the runtimes directory: ${content}`);
            }
        }
        catch (error) {
            logger_1.logger.error(`Error processing file ${filePath}:`, error);
        }
    }
    handleRemovedRuntime(filePath) {
        const fileName = path_1.default.basename(filePath);
        if (fileName in this.filenameToRuntimeMap) {
            const runtime = this.filenameToRuntimeMap[fileName];
            delete this.filenameToRuntimeMap[fileName];
            delete this.idToFileMap[runtime.id];
            this.eventEmitter.emit(types_1.RuntimeEvent.REMOVE, runtime);
            logger_1.logger.debug(`Removed runtime with id ${runtime.id}.`);
        }
    }
    httpErrorHandler(error, message) {
        const newError = new types_1.GenkitToolsError(message || 'Internal Error');
        if (error.response) {
            newError.data = error.response.data;
            newError.stack = (error.response?.data).message;
            if ((error.response?.data).message) {
                newError.data.data = {
                    ...newError.data.data,
                    genkitErrorMessage: message,
                    genkitErrorDetails: {
                        stack: (error.response?.data).message,
                        traceId: (error.response?.data).traceId,
                    },
                };
            }
            throw newError;
        }
        throw new types_1.GenkitToolsError(message || 'Internal Error', {
            cause: error.cause,
        });
    }
    async handleStreamError(error, message) {
        if (error.response &&
            error.config?.responseType === 'stream' &&
            error.response.data.on) {
            try {
                const body = await this.streamToString(error.response.data);
                try {
                    error.response.data = JSON.parse(body);
                }
                catch (e) {
                    error.response.data = {
                        message: body || 'Unknown error',
                    };
                }
            }
            catch (e) {
                error.response.data = {
                    message: 'Failed to read error response stream',
                    details: String(e),
                };
            }
        }
        this.httpErrorHandler(error, message);
    }
    streamToString(stream) {
        return new Promise((resolve, reject) => {
            let buffer = '';
            stream.on('data', (chunk) => {
                buffer += chunk.toString();
            });
            stream.on('end', () => {
                resolve(buffer);
            });
            stream.on('error', (err) => {
                reject(err);
            });
        });
    }
    async performHealthChecks() {
        const healthCheckPromises = Object.entries(this.filenameToRuntimeMap).map(async ([fileName, runtime]) => {
            if (!(await (0, utils_1.checkServerHealth)(runtime.reflectionServerUrl, runtime.id))) {
                await this.removeRuntime(fileName);
            }
        });
        return Promise.all(healthCheckPromises);
    }
    async removeRuntime(fileName) {
        try {
            const runtimesDir = await (0, utils_1.findRuntimesDir)(this.projectRoot);
            const runtimeFilePath = path_1.default.join(runtimesDir, fileName);
            await promises_1.default.unlink(runtimeFilePath);
        }
        catch (error) {
            logger_1.logger.debug(`Failed to delete runtime file: ${error}`);
        }
        logger_1.logger.debug(`Removed unhealthy runtime ${fileName} from manager.`);
    }
}
exports.RuntimeManager = RuntimeManager;
function isValidRuntimeInfo(data) {
    let timestamp = '';
    if (typeof data.timestamp === 'string') {
        timestamp = data.timestamp.replaceAll('_', ':');
    }
    return (typeof data === 'object' &&
        data !== null &&
        typeof data.id === 'string' &&
        typeof data.pid === 'number' &&
        typeof data.reflectionServerUrl === 'string' &&
        typeof data.timestamp === 'string' &&
        !isNaN(Date.parse(timestamp)) &&
        (data.name === undefined || typeof data.name === 'string'));
}
